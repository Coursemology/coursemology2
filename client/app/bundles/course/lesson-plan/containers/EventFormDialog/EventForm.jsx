import PropTypes from 'prop-types';
import { defineMessages, FormattedMessage } from 'react-intl';
import { reduxForm, Field, Form } from 'redux-form';
import renderTextField from 'lib/components/redux-form/TextField';
import RichTextField from 'lib/components/redux-form/RichTextField';
import renderAutoCompleteField from 'lib/components/redux-form/AutoComplete';
import Toggle from 'lib/components/redux-form/Toggle';
import DateTimePicker from 'lib/components/redux-form/DateTimePicker';
import formTranslations from 'lib/translations/form';
import translations from 'course/lesson-plan/translations';
import { formNames, fields } from 'course/lesson-plan/constants';

const {
  TITLE,
  EVENT_TYPE,
  LOCATION,
  DESCRIPTION,
  START_AT,
  END_AT,
  PUBLISHED,
} = fields;

const styles = {
  columns: {
    display: 'flex',
  },
  oneColumn: {
    flex: 1,
  },
  eventType: {
    flex: 1,
    marginRight: 10,
  },
  toggle: {
    marginTop: 16,
  },
};

const validationTranslations = defineMessages({
  startEndValidationError: {
    id: 'course.lessonPlan.EventForm.startEndValidationError',
    defaultMessage: 'Must be after "Start At"',
  },
});

const validate = (values) => {
  const errors = {};

  const requiredFields = ['title', 'start_at', 'event_type'];
  requiredFields.forEach((field) => {
    if (!values[field]) {
      errors[field] = formTranslations.required;
    }
  });

  if (
    values.end_at &&
    values.start_at &&
    new Date(values.start_at) >= new Date(values.end_at)
  ) {
    errors.end_at = validationTranslations.startEndValidationError;
  }

  return errors;
};

const EventForm = ({
  handleSubmit,
  onSubmit,
  disabled,
  formValues,
  shiftEndDate,
  eventTypes,
  eventLocations,
}) => (
  <Form onSubmit={handleSubmit(onSubmit)}>
    <Field
      fullWidth
      name="title"
      label={<FormattedMessage {...translations[TITLE]} />}
      component={renderTextField}
      {...{ disabled }}
    />
    <div style={styles.columns}>
      <Field
        fullWidth
        selectOnFocus
        name="event_type"
        label={<FormattedMessage {...translations[EVENT_TYPE]} />}
        component={renderAutoCompleteField}
        options={eventTypes}
        style={styles.eventType}
        {...{ disabled }}
      />
      <Field
        fullWidth
        selectOnFocus
        name="location"
        label={<FormattedMessage {...translations[LOCATION]} />}
        component={renderAutoCompleteField}
        options={eventLocations}
        style={styles.oneColumn}
        {...{ disabled }}
      />
    </div>
    <Field
      fullWidth
      name="description"
      label={<FormattedMessage {...translations[DESCRIPTION]} />}
      component={RichTextField}
      multiline
      rows={2}
      {...{ disabled }}
    />
    <div style={styles.columns}>
      <Field
        name="start_at"
        label={<FormattedMessage {...translations[START_AT]} />}
        component={DateTimePicker}
        afterChange={(_, newStartAt) =>
          shiftEndDate(formNames.EVENT, newStartAt, formValues)
        }
        style={styles.oneColumn}
        {...{ disabled }}
      />
      <Field
        name="end_at"
        label={<FormattedMessage {...translations[END_AT]} />}
        component={DateTimePicker}
        style={styles.oneColumn}
        {...{ disabled }}
      />
    </div>
    <Field
      name="published"
      component={Toggle}
      parse={Boolean}
      label={<FormattedMessage {...translations[PUBLISHED]} />}
      style={styles.toggle}
      disabled={disabled}
    />
  </Form>
);

EventForm.propTypes = {
  eventTypes: PropTypes.arrayOf(PropTypes.string),
  eventLocations: PropTypes.arrayOf(PropTypes.string),
  formValues: PropTypes.shape({}),
  shiftEndDate: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
};

export default reduxForm({
  form: formNames.EVENT,
  validate,
})(EventForm);
