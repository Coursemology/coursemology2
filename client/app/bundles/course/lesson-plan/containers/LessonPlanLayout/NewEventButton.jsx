import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  injectIntl,
  defineMessages,
  intlShape,
  FormattedMessage,
} from 'react-intl';
import { Button } from '@material-ui/core';
import { showEventForm, createEvent } from 'course/lesson-plan/actions';

const translations = defineMessages({
  newEvent: {
    id: 'course.lessonPlan.NewEventButton.newEvent',
    defaultMessage: 'New Event',
  },
  success: {
    id: 'course.lessonPlan.NewEventButton.success',
    defaultMessage: 'Event created.',
  },
  failure: {
    id: 'course.lessonPlan.NewEventButton.failure',
    defaultMessage: 'Failed to create event.',
  },
});

class NewEventButton extends Component {
  createEventHandler = (data) => {
    const { dispatch } = this.props;
    const successMessage = <FormattedMessage {...translations.success} />;
    const failureMessage = <FormattedMessage {...translations.failure} />;
    return dispatch(createEvent(data, successMessage, failureMessage));
  };

  showForm = () => {
    const { dispatch, intl } = this.props;
    return dispatch(
      showEventForm({
        onSubmit: this.createEventHandler,
        formTitle: intl.formatMessage(translations.newEvent),
        initialValues: {},
      }),
    );
  };

  render() {
    if (!this.props.canManageLessonPlan) {
      return <div />;
    }

    return (
      <Button variant="contained" color="primary" onClick={this.showForm}>
        <FormattedMessage {...translations.newEvent} />
      </Button>
    );
  }
}

NewEventButton.propTypes = {
  canManageLessonPlan: PropTypes.bool.isRequired,

  dispatch: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

export default connect((state) => ({
  canManageLessonPlan: state.flags.canManageLessonPlan,
}))(injectIntl(NewEventButton));
