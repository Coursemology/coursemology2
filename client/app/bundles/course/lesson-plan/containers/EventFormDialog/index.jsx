import { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import FormDialogue from 'lib/components/form/FormDialogue';

import { actions } from '../../store';

import EventForm from './EventForm';

const EventFormDialog = ({
  visible,
  disabled,
  formTitle,
  initialValues,
  onSubmit,
  dispatch,
  items,
}) => {
  const [isDirty, setIsDirty] = useState(false);

  const { eventTypes, eventLocations } = items.reduce(
    (values, item) => {
      if (!item.eventId) {
        return values;
      }
      if (item.location) {
        values.eventLocations.push(item.location);
      }
      values.eventTypes.push(item.lesson_plan_item_type[0]);
      return values;
    },
    { eventTypes: [], eventLocations: [] },
  );

  return (
    <FormDialogue
      disabled={disabled}
      form="event-form"
      hideForm={() => dispatch(actions.hideEventForm())}
      open={visible}
      skipConfirmation={!isDirty}
      title={formTitle}
    >
      <EventForm
        {...{ initialValues, onSubmit, disabled }}
        emitsVia={(eventForm) => setIsDirty(eventForm.isDirty)}
        eventLocations={[...new Set(eventLocations)]}
        eventTypes={[...new Set(eventTypes)]}
      />
    </FormDialogue>
  );
};

EventFormDialog.defaultProps = {
  visible: false,
  disabled: false,
};

EventFormDialog.propTypes = {
  visible: PropTypes.bool,
  disabled: PropTypes.bool,
  formTitle: PropTypes.string,
  initialValues: PropTypes.shape({
    id: PropTypes.number,
    eventId: PropTypes.number,
    title: PropTypes.string,
    event_type: PropTypes.string,
    location: PropTypes.string,
    description: PropTypes.string,
    start_at: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.instanceOf(Date),
    ]),
    end_at: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
    published: PropTypes.bool,
  }),
  items: PropTypes.arrayOf(
    PropTypes.shape({
      eventId: PropTypes.number,
      location: PropTypes.string,
      lesson_plan_item_type: PropTypes.arrayOf(PropTypes.string),
    }),
  ),
  onSubmit: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default connect(({ lessonPlan }) => ({
  ...lessonPlan.eventForm,
  items: lessonPlan.lessonPlan.items,
}))(EventFormDialog);
