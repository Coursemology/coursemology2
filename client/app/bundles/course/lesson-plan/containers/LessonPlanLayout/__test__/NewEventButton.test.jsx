import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils';
import CourseAPI from 'api/course';
import storeCreator from 'course/lesson-plan/store';
import EventFormDialog from 'course/lesson-plan/containers/EventFormDialog';
import NewEventButton from '../NewEventButton';

describe('<NewEventButton />', () => {
  it('allows event to be created via EventFormDialog', () => {
    const spyCreate = jest.spyOn(CourseAPI.lessonPlan, 'createEvent');
    const store = storeCreator({ flags: { canManageLessonPlan: true } });
    const contextOptions = buildContextOptions(store);

    const eventFormDialog = mount(<EventFormDialog />, contextOptions);
    const newEventButton = mount(<NewEventButton />, contextOptions);

    // Click 'new event' button
    newEventButton.find('button').simulate('click');
    expect(
      eventFormDialog.update().find('EventFormDialog').first().props().visible,
    ).toBe(true);

    // Fill event form
    const eventData = {
      title: 'Ambitious event title',
      event_type: 'In-person Meetup',
      start_at: new Date('2016-12-31T16:00:00.000Z'),
    };
    const startAt = '01-01-2017';
    const eventForm = eventFormDialog.find('form');
    const titleInput = eventForm.find('input[name="title"]');
    titleInput.simulate('change', { target: { value: eventData.title } });
    const eventTypeInput = eventForm.find('input[name="event_type"]');
    eventTypeInput.simulate('change', {
      target: { value: eventData.event_type },
    });
    eventTypeInput.simulate('blur');
    const startAtDateInput = eventForm.find('input[name="start_at"]').first();
    startAtDateInput.simulate('change', { target: { value: startAt } });
    startAtDateInput.simulate('blur');

    // Submit event form
    const submitButton = eventFormDialog
      .find('FormDialogue')
      .first()
      .instance().submitButton;
    ReactTestUtils.Simulate.click(ReactDOM.findDOMNode(submitButton));
    expect(spyCreate).toHaveBeenCalledWith({ lesson_plan_event: eventData });
  });

  it('is hidden when canManageLessonPlan is false', () => {
    const store = storeCreator({ flags: { canManageLessonPlan: false } });
    const contextOptions = buildContextOptions(store);
    const newEventButton = mount(<NewEventButton />, contextOptions);
    expect(newEventButton).toMatchSnapshot();
  });
});
