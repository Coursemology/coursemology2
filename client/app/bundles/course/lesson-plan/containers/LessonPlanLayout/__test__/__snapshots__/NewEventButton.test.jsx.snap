// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewEventButton /> is hidden when canManageLessonPlan is false 1`] = `
<Connect(injectIntl(NewEventButton))>
  <injectIntl(NewEventButton)
    canManageLessonPlan={false}
    dispatch={[Function]}
  >
    <NewEventButton
      canManageLessonPlan={false}
      dispatch={[Function]}
      intl={
        {
          "$t": [Function],
          "defaultFormats": {},
          "defaultLocale": "en",
          "defaultRichTextElements": undefined,
          "fallbackOnEmptyString": true,
          "formatDate": [Function],
          "formatDateTimeRange": [Function],
          "formatDateToParts": [Function],
          "formatDisplayName": [Function],
          "formatList": [Function],
          "formatListToParts": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatNumberToParts": [Function],
          "formatPlural": [Function],
          "formatRelativeTime": [Function],
          "formatTime": [Function],
          "formatTimeToParts": [Function],
          "formats": {},
          "formatters": {
            "getDateTimeFormat": [Function],
            "getDisplayNames": [Function],
            "getListFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralRules": [Function],
            "getRelativeTimeFormat": [Function],
          },
          "locale": "en",
          "messages": {},
          "onError": [Function],
          "onWarn": [Function],
          "textComponent": Symbol(react.fragment),
          "timeZone": "Asia/Singapore",
          "wrapRichTextChunksInFragment": undefined,
        }
      }
    >
      <div />
    </NewEventButton>
  </injectIntl(NewEventButton)>
</Connect(injectIntl(NewEventButton))>
`;
